src/ngraph_builder.cc

```diff
@@ -1351,6 +1351,126 @@ static Status TranslateConv2DBackpropInputOp(
+    // NHWC
+    case 0:
+      channel_dimension = 3;
+      break;
+    // NCHW or NCHW_VEC_C
+    case 1:
+      channel_dimension = 1;
+      break;
+    default:
+      return errors::InvalidArgument(
+          "DepthToSpace supported data format is NCHW, NHWC, or NCHW_VECT_C");
+  }
+
+  // Error checking : depth must be divisible by square of the block_size
+  if (input_shape[channel_dimension] % (block_size * block_size) != 0) {
+    return errors::InvalidArgument(
+        "Input tensor's channel dimension ,", input_shape[channel_dimension],
+        " is not divisible by square of the block_size ", block_size);
+  }
+
+  ng::AxisVector ng_reshape_shape;
+  ng::AxisVector ng_transpose_shape;
+  ng::AxisVector ng_output_shape;
+
+  switch (format_to_int_map[tf_data_format]) {
+    // NHWC
+    case 0: {
+      ng_reshape_shape.push_back(input_shape[0]);  // N dimension
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_reshape_shape.push_back(input_shape[i + 1]);
+      }
+      int64 num_blocks = 1;
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_reshape_shape.push_back(block_size);
+        num_blocks *= block_size;
+      }
+      ng_reshape_shape.push_back(input_shape[channel_dimension] / num_blocks);
+
+      ng_transpose_shape.push_back(0);
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_transpose_shape.push_back(i + 1);
+        ng_transpose_shape.push_back(i + 1 + num_spatial_dimensions);
+      }
+      ng_transpose_shape.push_back(channel_dimension + num_spatial_dimensions);
+
+      ng_output_shape.push_back(input_shape[0]);
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_output_shape.push_back(input_shape[i + 1] * block_size);
+      }
+      ng_output_shape.push_back(input_shape[channel_dimension] / num_blocks);
+      break;
+    }  // end of case NHWC
+    // NCHW
+    case 1: {
+      int64 num_blocks = 1;
+      ng_reshape_shape.push_back(input_shape[0]);  // N dimension
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_reshape_shape.push_back(block_size);
+        num_blocks *= block_size;
+      }
+      ng_reshape_shape.push_back(input_shape[channel_dimension] / num_blocks);
+
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_reshape_shape.push_back(input_shape[i + 2]);
+      }
+
+      ng_transpose_shape.push_back(0);
+      ng_transpose_shape.push_back(1 + num_spatial_dimensions);
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_transpose_shape.push_back(i + 2 + num_spatial_dimensions);
+        ng_transpose_shape.push_back(i + 1);
+      }
+
+      ng_output_shape.push_back(input_shape[0]);
+      ng_output_shape.push_back(input_shape[channel_dimension] / num_blocks);
+      for (int i = 0; i < num_spatial_dimensions; i++) {
+        ng_output_shape.push_back(input_shape[i + 2] * block_size);
+      }
+      break;
+    }  // end of case NCHW
+  }
+
```

Comment:
As discussed, we can have the logic for calculating reshape, transpose, output shapes for only one of the data format NCHW (or NHWC), which we can assume as the standard format. If input has the other format, we reshape the input to the std format , do our computations and reshape it back.  Please try this.

We will have a simpler implementation and bug fixes (later if any) would be simpler to introduce. We only have to worry computation in one format.

Reply:
Hi @shresthamalik @sayantan-nervana, I discussed with @avijit-nervana and @aprocter today, we agreed on that the each op in the bridge should honor its' input tensor's data format, so it is ok or maybe more clearer to keep the logic separate in the translation function. In addition, adding transpose/reshape at the beginning and the end for another format will introduce two additional operations, it might be not good in terms of performance concern.

What do you think? Feel free to input your thoughts.

Reply:
Ok, then we can keep this implementation to save the additional Reshape ops.